src/modules/aws_ec2/mappers/instance_block_device_mapping.ts

```diff
@@ -0,0 +1,282 @@
+
+        // only attach if volume is not in use. We can have the case that the volume is auto-attached on creation
+        if (e.volume.state === VolumeState.AVAILABLE) {
+          await this.attachVolume(client.ec2client, e.volume.volumeId, e.instance.instanceId, e.deviceName);
+        }
+
+        // return mapping
+        const newMap: InstanceBlockDeviceMapping = {
+          deviceName: e.deviceName,
+          instanceId: e.instanceId,
+          instance: e.instance,
+          volumeId: e.volumeId,
+          volume: e.volume,
+          region: e.instance.region,
+          deleteOnTermination: e.deleteOnTermination,
+        };
+
+        // Save the record back into the database to get the new fields updated
+        newMap.id = e.id;
+        await this.module.instanceBlockDeviceMapping.db.update(newMap, ctx);
+        out.push(newMap);
+      }
+      return out;
+    },
+    read: async (ctx: Context, id?: string) => {
+      if (id) {
+        // decompose the id
+        const { instanceId, volumeId, region } = this.idFields(id);
+        const client = (await ctx.getAwsClient(region)) as AWS;
+
+        // check if we can find the instance in database
+        const instance = await this.module.instance.db.read(
+          ctx,
+          this.module.instance.generateId({ instanceId: instanceId ?? '', region }),
+        );
+
+        if (instance) {
+          // read the instance mapping
+          const mapping = await this.module.instance.getInstanceBlockDeviceMapping(
+            client.ec2client,
+            instanceId,
+          );
+          for (const newMap of mapping ?? []) {
+            if (newMap.DeviceName && newMap.Ebs?.VolumeId === volumeId) {
+              const res = await this.instanceBlockDeviceMappingMapper(newMap, instance, ctx);
+              return res;
+            }
+          }
+        }
+      } else {
+        const out: InstanceBlockDeviceMapping[] = [];
+        const enabledRegions = (await ctx.getEnabledAwsRegions()) as string[];
+        await Promise.all(
+          enabledRegions.map(async region => {
+            const client = (await ctx.getAwsClient(region)) as AWS;
+            const rawInstances = (await this.module.instance.getInstances(client.ec2client)) ?? [];
+            for (const i of rawInstances) {
+              // exclude spot instances and terminating ones
+              if (i.InstanceLifecycle === InstanceLifecycle.SPOT) continue;
+              if (i.State?.Name === 'terminated' || i.State?.Name === 'shutting-down') continue;
+              if (!i.InstanceId) continue; // if we do not have an id we skip
+
+              // check instance block device mappings
+              const mapping = await this.module.instance.getInstanceBlockDeviceMapping(
+                client.ec2client,
+                i.InstanceId,
+              );
+
+              // check if we can find the instance in database
+              const instance = await this.module.instance.db.read(
```

Comment:
Depending on the `db.read` path in a `cloud.read` function feels like the wrong choice.